src/main/java/sg/ncl/MainController.java

```diff
@@ -2105,113 +2101,118 @@ public String data(Model model, HttpSession session) throws Exception {
     	return "data";
     }
 
-//    @RequestMapping(value="/data/contribute", method=RequestMethod.GET)
-//    public String contributeData(Model model) {
-//    	model.addAttribute("dataset", new Dataset());
-//
-//    	File rootFolder = new File(App.ROOT);
-//    	List<String> fileNames = Arrays.stream(rootFolder.listFiles())
-//    			.map(f -> f.getError())
-//    			.collect(Collectors.toList());
-//
-//    		model.addAttribute("files",
-//    			Arrays.stream(rootFolder.listFiles())
-//    					.sorted(Comparator.comparingLong(f -> -1 * f.lastModified()))
-//    					.map(f -> f.getError())
-//    					.collect(Collectors.toList())
-//    		);
-//
-//    	return "contribute_data";
-//    }
+    @RequestMapping(value={"/data/contribute", "/data/contribute/{id}"}, method=RequestMethod.GET)
+    public String contributeData(Model model, @PathVariable Optional<String> id) throws Exception {
+        if (id.isPresent()) {
+            HttpEntity<String> request = createHttpEntityHeaderOnly();
+            ResponseEntity response = restTemplate.exchange(properties.getDataset(id.get()), HttpMethod.GET, request, String.class);
+            String dataResponseBody = response.getBody().toString();
+            JSONObject dataInfoObject = new JSONObject(dataResponseBody);
+            Dataset dataset = extractDataInfo(dataInfoObject.toString());
+            model.addAttribute("dataset", dataset);
+        } else {
+            model.addAttribute("dataset", new Dataset());
+        }
+    	return "data_contribute";
```

Comment:
![CRITICAL](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-critical.png) Define a constant instead of duplicating this literal "data_contribute" 3 times. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1192)